content/blog/_posts/2021-05-java-client.md

```diff
@@ -0,0 +1,660 @@
+`lib/src/test/java/my/java/lib/FooTest.java`:
+
+```java
+// lib/src/test/java/my/java/lib/FooTest.java
+package my.java.lib;
+
+import org.junit.Test;
+import static org.junit.Assert.*;
+import my.java.lib.Foo;
+
+public class FooTest {
+    @Test public void testSomeLibraryMethod() {
+        Foo foo = new Foo(10);
+        assertTrue("Foo.val", foo.val() == 15); // This will fail later.
+    }
+}
+```
+
+Now if you run `./gradlew test` you should get the following error:
+
+```bash
+$ ./gradlew test
+
+> Task :lib:test FAILED
+
+my.java.lib.FooTest > testSomeLibraryMethod FAILED
+    java.lang.UnsatisfiedLinkError at FooTest.java:9
+
+2 tests completed, 1 failed
+
+FAILURE: Build failed with an exception.
+
+* What went wrong:
+Execution failed for task ':lib:test'.
+> There were failing tests. See the report at: file:///home/simlay/projects/infinyon/fluvio-demo-apps-rust/my-java-lib-blog-post/lib/build/reports/tests/test/index.html
+
+* Try:
+Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
+
+* Get more help at https://help.gradle.org
+
+BUILD FAILED in 741ms
+3 actionable tasks: 2 executed, 1 up-to-date
+```
+
+This means we need to link our rust as a static library.
+
+## Gradle build
+
+The `gradle init` step in the [setup](#setup) generated a project template. We need to add the project template to `lib/build.gradle`:
+
+```groovy
+// Append to `lib/build.gradle`
+def rustBasePath = ".."
+
+// execute cargo metadata and get path to target directory
+tasks.create(name: "cargo-output-dir", description: "Get cargo metadata") {
+    new ByteArrayOutputStream().withStream { os ->
+        exec {
+            commandLine 'cargo', 'metadata', '--format-version', '1'
+            workingDir rustBasePath
+            standardOutput = os
+        }
+        def outputAsString = os.toString()
+        def json = new groovy.json.JsonSlurper().parseText(outputAsString)
+
+        logger.info("cargo target directory: ${json.target_directory}")
+        project.ext.cargo_target_directory = json.target_directory
+    }
+}
+
+// Build with cargo
+tasks.create(name: "cargo-build", type: Exec, description: "Running Cargo build", dependsOn: "cargo-output-dir") {
+    workingDir rustBasePath
+    commandLine 'cargo', 'build', '--release'
+}
+
+tasks.create(name: "rust-deploy", type: Sync, dependsOn: "cargo-build") {
+    from "${project.ext.cargo_target_directory}/release"
+    include "*.dylib","*.so"
+    into "rust-lib/"
+}
+
+clean.dependsOn "clean-rust"
+tasks.withType(JavaCompile) {
+    compileTask -> compileTask.dependsOn "rust-deploy"
+}
+
+sourceSets {
+    main {
+        java {
+            srcDir 'src/main/java'
+        }
+        resources {
+            srcDir 'rust-lib'
+        }
+    }
+}
+```
+
+In short, this adds:
```

Comment:
"Adds" doesn't flow well into the bullets
- Adds runs
- Adds copies
- Adds and then adds